### PR TITLE
set up show and edit pages for lessons

### PR DESCRIPTION
**Pull request by davidsbailey**
The goal of this PR is to propose (and implement) a skeleton for how to handle lesson (aka lesson plan) editing and displaying in our new system. The assumptions baked into this specific proposal are:

1. **do not worry about the final URL format for now**. instead, use the url format that's easiest to do with rails, which is to reference each lesson by its unique id. this value is not consistent across environments, but this should be ok for editing / viewing. this is the model used by levels, which are viewed / edited by ids (when iterating on levelbuilder), but not scripts or courses, which are viewed / edited by name. Before the January launch deadline, we should revisit this decision and decide if we want to use a different url such as /s/script_name/stage/1 (to match existing urls), or perhaps something more ambitious.
2. **do not worry about implementing serialize / seed in this PR**, but do set us up so that someone (@uponthesun ) can work on that while someone else (@dmcavoy ) works on the show and edit pages for lessons, and/or possibly the calendar view that goes on the unit overview page ([example](https://curriculum.code.org/csd-20/unit2/)).
3. **only allow editing of "new" fields on the lesson edit page**. that means that editing which levels are in a level edit page will keep living on the script edit page in the DSL text area for now. "display name" and "key" will also not be editable on the new page.
4. **provide a way to get to the lesson edit page via the script edit gui**, which seems slightly easier for developers than having to copy IDs around. this gives us a way to get to this page, without exposing it to curriculum writers, without new experiment code. this is a bit hacky in the sense that the script edit gui page doesn't work for saving right now, but can be used in this one specific way to get to the lesson edit page for development purposes.

Feedback on this approach is quite welcome! I tried to pick a stopping point which proves out what this will look like, without having done so much work that I'd be too sad to undo some of it.

FWIW, This PR is interesting in the sense that I got to use a bunch of rails conventions (and other shortcuts like test_user_gets_response_for) that we haven't necessarily been using that consistently. please feel free to comment even if you are just wondering how any of the rails magic is working here.

## Future work

in a next PR, I'll move the edit form into React so that any additions to it take us in the right technical direction.

## Testing story

Unit tests covering new lesson routes. Manual testing as follows:

edit-lesson

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/views/lessons/show.html.haml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+%h1= @lesson.name
```
Comment (dmcavoy) [inline dashboard/app/views/lessons/show.html.haml:1]: Are you going to set this up into react too or is that follow up work we should plan to tackle later?

File: dashboard/config/routes.rb [hunk 2]
```diff
@@ -309,6 +309,8 @@
   resources :courses, param: 'course_name'
   get '/course/:course_name', to: redirect('/courses/%{course_name}')
 
+  resources :lessons, only: [:show, :edit, :update]
```
Comment (dmcavoy) [inline dashboard/config/routes.rb:312]: I would love a link to read about how this works.

Comment (davidsbailey) [reply dashboard/config/routes.rb:312]: here is a good starting point: https://guides.rubyonrails.org/routing.html#resource-routing-the-rails-default

let me know if you have follow-up questions.

File: dashboard/app/views/lessons/show.html.haml (hunk 1 above)
Comment (davidsbailey) [reply dashboard/app/views/lessons/show.html.haml:1]: this should get converted into react. I would be happy to tackle that too in the next PR.

File: dashboard/test/controllers/lessons_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    get :show, params: {
+      id: @lesson.id
+    }
+    assert_response :ok
+    assert(@response.body.include?(@lesson.name))
+    assert(@response.body.include?(@lesson.overview))
+  end
+
+  # only levelbuilders can edit
+  test_user_gets_response_for :edit, params: -> {{id: @lesson.id}}, user: nil, response: :redirect
```
Comment (dmcavoy) [inline dashboard/test/controllers/lessons_controller_test.rb:41]: where does this redirect to?

**Review by dmcavoy — APPROVED** [review]
This approach looks good to me. Thanks for getting this set up.

File: dashboard/test/controllers/lessons_controller_test.rb (hunk 3 above)
Comment (davidsbailey) [reply dashboard/test/controllers/lessons_controller_test.rb:41]: to the login page.

File: dashboard/config/routes.rb (hunk 2 above)
Comment (davidsbailey) [reply dashboard/config/routes.rb:312]: the other mysterious thing to read about in this PR is CanCan. https://github.com/CanCanCommunity/cancancan it has to do with the following aspects of this PR:
* `load_and_authorize_resource`
* `can?`
* ability.rb

File: dashboard/app/views/lessons/show.html.haml (hunk 1 above)
Comment (davidsbailey) [reply dashboard/app/views/lessons/show.html.haml:1]: https://github.com/code-dot-org/code-dot-org/pull/36392

File: dashboard/test/controllers/lessons_controller_test.rb (hunk 3 above)
Comment (davidsbailey) [reply dashboard/test/controllers/lessons_controller_test.rb:41]: I've opened a follow up PR to make it easier to test this: https://github.com/code-dot-org/code-dot-org/pull/36391 .

File: dashboard/config/routes.rb (hunk 2 above)
Comment (dmcavoy) [reply dashboard/config/routes.rb:312]: I will look into these both a little bit more

**Pull request merged**